### DISCUSSION TITLE
Add software transition option to the Light entity?

### DISCUSSION THREAD

**Discussion by emichael** [home-assistant/architecture#585] [Entity Models]
Currently, Home Assistant only executes transitions for light.turn_on etc. if the underlying integration supports them. Even if the device doesn't support transitions for brightness, color, etc., they could be done in software. The Light entity could provide a default implementation of this. (I have some ideas on the right way to do the software transition I'd be happy to discuss.) Lots of people have done this in scripts/Python services, etc., but doing this at the root Light entity would obviously be the most elegant.

I haven't seen much discussion of the idea, so I thought I'd raise it here to see if there's any possibility of this happening at the Light entity level. It's a big change, but if HA provided a way to opt-in to software transition on a per-light basis, then maybe there's a way to do this gracefully? I'm still learning my way around the HA codebase, so I don't have much to say about that myself.

Sorry in advance if this is the wrong place to raise this.

**Comment by frenck** [comment]
It is not our goal or task to simulate things. If a device doesn't support transitions, we should not simulate those; is has implications for the state machine (lots of state updates, although could be worked around) and above all, not all devices will be able to just handle that (so it becomes a case-by-case implementation).

Basically, simulating will end up sending a lot of state changes.

**Reply by emichael** [reply]
Correct me if I'm wrong, but some device-generated transitions also create lots of state updates, don't they? I can watch my LIFX bulb's brightness fade in the dashboard, for instance.

I agree that it's never going to be the perfect. As you say, some lights definitely won't be able to handle updates at a rate that makes a reasonably smooth transition. But if it's opt-in, then individual integrations and users could try a software transition, and if it doesn't work, disable it again.

It's not simulating, though. It's sending a series of commands to the bulb that the bulb accepts and executes. Lots of people have written various scripts to do it, they just don't work nearly as well as something that runs at the entity level. You have to do any of a number of hacky things to try to detect manual changes to the light during the fade and abort, for instance.

**Comment by frenck** [comment]
> But if it's opt-in, then individual integrations and users could try a software transition, and if it doesn't work, disable it again.

It is simulating, you are suggesting we implement a feature that does not exist on the target device. We should not emulate/simulate.

**Reply by emichael** [reply]
When you put it that way, isn't that the whole point of HA? My light doesn't have the ability to detect my geolocation on my phone and turn on when I come home, for example.

I guess the definition of "simulate" doesn't really matter, but I am genuinely confused. Why is it a problem for HA to try to enhance the capabilities of various smart things? And why is this specific suggestion unworkable? You say that it's possible to work around the issue of the light generating excessive state updates; you can probably just rate-limit the update events emitted? I was thinking that you'd probably want to rate-limit the commands sent to the lights anyways, to at most 1 Hz.

**Reply by frenck** [reply]
> When you put it that way, isn't that the whole point of HA?

? I don't think using these arguments and comments is needed or fitting in an architectural discussion.

> Why is it a problem for HA to try to enhance the capabilities of various smart things? And why is this specific suggestion unworkable?

Its device-specific. Some devices may choke if you hit them often to simulate a transition, e.g., not able to handle updates fast enough, overloading their protocols and sometimes even their mesh networks. Thus is a risk to put out such a feature in general.

Implementing device features in software is a general problem, we had a lot of cases that went south, e.g., simulating cover/gates opening/closing states and many other examples.

**Reply by thecode** [reply]
As I see it this is integration specific, if a device can support transition by simulating it with software than an integration can implement it (and it is preferred to implement it in the external device library and not inside HA core). There should not be a core component that will do it without the device specific implementation at the integration side.

**Reply by emichael** [reply]
> ? I don't think using these arguments and comments is needed or fitting in an architectural discussion.

I was just trying to elucidate what I didn't understand about the point you were making.

I completely agree that it's device/integration specific. Turning on such a feature by default would be a terrible idea. My point is that lots of people are doing it *already*, just poorly. I myself wrote a terribly hacky pyscript thing that sort of works, but it's also about the best you can do at that level.

While individual integrations could certainly have their own software faders (or call one via an external library), it's actually not a totally trivial thing to do. All I'm suggesting is that HA provide a default implementation that could get reused, to alleviate the burden from integration maintainers. If end-users could opt-in without relying on integrations getting individually updated, all the better.

**Reply by frenck** [reply]
So, additionally, it may work for some people and not for all. For example, a single Z-Wave bulb, might work and could be implemented for a specific device. But if there is a large network, such a working thing for 1, 2, maybe even 5 bulbs may work. Next, a user with a large network comes along, overloading his Z-Wave mesh network by such a feature.

This is the reason why such features are not accepted in general. It's unpredictable, unreliable and hard to debug for an end-user.

**Reply by amelchio** [reply]
I agree that this would always be unreliable. For another example, a fading light might turn back on if it was turned off by an external switch. We already have this problem with the Flux integration which does a scheduled fade.

Maybe the feature could be implemented as a custom component?

**Comment by frenck** [comment]
As per the above discussion and comments, we currently do not aim to allow such simulated features (as it may have additional repercussions).

If a user wants to do so themselves using a script (or custom integration) that is fine, as it becomes their responsibility at that point. From a Core perspective, it is currently better to stick with representing the device as natively as possible.

../Frenck
